File: c/binary_search.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

const char NUMS[] = {-5, -3, 0, 1, 3, 8, 12};

int search_binary(int target) {
  int low = 0;
  int high = strlen(NUMS);

  while (low <= high) {
    int mid = (low + high) / 2;

    if (NUMS[mid] == target) {
      return mid;
    } else if (NUMS[mid] < target) {
      low = mid + 1;
    } else {
      high = mid -1;
    }
  }


  return -1;
}

int main(int argc, char *argv[]) {
  if (argc < 2) {
    return 1;
  } else {
    char *endptr;
    long target = strtol(argv[1], &endptr, 10);

    if (*endptr != '\0') {
      printf("invalid");
      return 1;
    }

    search_binary(target);
  }
  return 0;
}

File: rust/src/binary_search.rs
const _NUMS: [i32; 7] = [-5, -3, 0, 1, 3, 8, 12];

pub fn _binary_search(target: i32) -> i32 {
    let mut low = 0;
    let mut high = _NUMS.len() - 1;

    while low <= high {
        let mid = (low + high) / 2;

        if _NUMS[mid] == target {
            return mid as i32;
        } else if _NUMS[mid] < target {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }

    -1
}
